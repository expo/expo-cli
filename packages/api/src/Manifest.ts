import { JSONObject } from '@expo/json-file';

import ApiV2, { ApiV2ClientOptions } from './ApiV2';
import UserManager from './UserManager';

export async function signAsync(
  user: ApiV2ClientOptions | null,
  manifest: JSONObject
): Promise<string> {
  const { signature } = await ApiV2.clientForUser(user).postAsync('manifest/eas/sign', {
    manifest,
  });
  return signature;
}

export async function signLegacyAsync(
  user: ApiV2ClientOptions | null,
  manifest: JSONObject
): Promise<string> {
  const { response } = await ApiV2.clientForUser(user).postAsync('manifest/sign', {
    args: {
      remoteUsername: manifest.owner ?? (await UserManager.getCurrentUsernameAsync()),
      remotePackageName: manifest.slug,
    },
    manifest: manifest as JSONObject,
  });
  return response;
}
