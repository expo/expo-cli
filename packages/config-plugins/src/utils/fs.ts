import fs from 'fs';
import path from 'path';

/** A basic function that copies a single file to another file location. */
export async function copyFilePathToPathAsync(src: string, dest: string): Promise<void> {
  const srcFile = await fs.promises.readFile(src);
  await fs.promises.mkdir(path.dirname(dest), { recursive: true });
  await fs.promises.writeFile(dest, srcFile);
}
