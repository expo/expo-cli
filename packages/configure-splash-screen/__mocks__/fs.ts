'use strict';
import { fs } from 'memfs';
module.exports = fs;
