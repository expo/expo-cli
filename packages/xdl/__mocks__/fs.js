const mockfs = require('mock-fs');

const container = {
  __internalFs: null,
  __configureFs: conf => {
    container.__internalFs = mockfs.fs(conf);
  },
  __mockFile: mockfs.file,
  __mockDirectory: mockfs.directory,
  __mockSymlink: mockfs.symlink,
};

const proxyfs = new Proxy(container, {
  get(target, property, receiver) {
    if (target.hasOwnProperty(property)) {
      return target[property];
    } else {
      if (target.__internalFs === null) {
        target.__internalFs = mockfs.fs();
      }

      return target.__internalFs[property];
    }
  },
});

module.exports = proxyfs;
