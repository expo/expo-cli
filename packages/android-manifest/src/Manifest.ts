import fs from 'fs-extra';
import { Attribute, Document, Element, parseXml } from 'libxmljs';
import { EOL } from 'os';
import path from 'path';

const USES_PERMISSION = 'uses-permission';

export function manipulateAttributeInTag(
  doc: Document,
  attribute: string,
  tag: string,
  manipulate: (attribute: Attribute) => any
): Array<any> {
  const elements = doc.find(`//${tag}`);
  return elements.map(element => {
    const targetAttribute = element.attr(attribute);
    if (targetAttribute) {
      return manipulate(targetAttribute);
    }
    return null;
  });
}

export function createPermission(doc: Document, name: string): Element {
  const permission = new Element(doc, USES_PERMISSION);
  permission.attr({ 'android:name': name });
  return permission;
}

export function removePermissions(doc: Document, permissionNames?: string[]) {
  const targetNames = permissionNames ? permissionNames.map(ensurePermissionNameFormat) : null;

  for (const attribute of getPermissionAttributes(doc)) {
    if (targetNames) {
      if (targetNames.includes(attribute.value())) {
        // continue to iterate in case there are duplicates
        attribute.remove();
      }
    } else {
      attribute.remove();
    }
  }
}

export function addPermission(doc: Document, permission: Element) {
  const elements: Element[] = doc.find(`//${USES_PERMISSION}`);

  if (elements.length) {
    return elements[elements.length - 1].addNextSibling(permission);
  }

  getRoot(doc).addChild(permission);
  return;
}

export function ensurePermissions(
  doc: Document,
  permissionNames: string[]
): { [permission: string]: boolean } {
  const permissions = getPermissions(doc);

  const results: { [permission: string]: boolean } = {};
  for (const permissionName of permissionNames) {
    const targetName = ensurePermissionNameFormat(permissionName);
    if (!permissions.includes(targetName)) {
      const permissionElement = createPermission(doc, targetName);
      addPermission(doc, permissionElement);
      results[permissionName] = true;
    } else {
      results[permissionName] = false;
    }
  }
  return results;
}

export function ensurePermission(doc: Document, permissionName: string): boolean {
  const permissions = getPermissions(doc);
  const targetName = ensurePermissionNameFormat(permissionName);

  if (!permissions.includes(targetName)) {
    const permissionElement = createPermission(doc, targetName);
    addPermission(doc, permissionElement);
    return true;
  }
  return false;
}

export function getRoot(doc: Document): Element {
  const root = doc.root();
  if (!root) throw new Error('no root');
  return root;
}

export function findAttributeInTag(doc: Document, attribute: string, tag: string): Attribute[] {
  const values: Array<Attribute | null> = manipulateAttributeInTag(
    doc,
    attribute,
    tag,
    (attribute: Attribute) => {
      return attribute;
    }
  );
  return values.filter(value => value !== null) as Attribute[];
}

export function ensurePermissionNameFormat(permissionName: string): string {
  if (permissionName.includes('.')) {
    const com = permissionName.split('.');
    const name = com.pop() as string;
    return [...com, name.toUpperCase()].join('.');
  } else {
    // If shorthand form like `WRITE_CONTACTS` is provided, expand it to `android.permission.WRITE_CONTACTS`.
    return ensurePermissionNameFormat(`android.permission.${permissionName}`);
  }
}

export function getPermissionAttributes(doc: Document): Attribute[] {
  const tags: Attribute[] = [];
  // newly added permissions are found with `android:name` instead of `name` (?)
  for (const id of ['name', 'android:name']) {
    tags.push(...findAttributeInTag(doc, id, 'uses-permission'));
  }
  return tags;
}

export function getPermissions(doc: Document): string[] {
  return getPermissionAttributes(doc).map(element => element.value());
}

export function logManifest(doc: Document) {
  console.log(getRoot(doc).toString());
}

const stringTimesN = (n: number, char: string) => Array(n + 1).join(char);

export function format(manifest: any, { indentLevel = 2, newline = EOL } = {}): string {
  let xmlInput: string;
  if (typeof manifest === 'string') {
    xmlInput = manifest;
  } else if (manifest.toString) {
    xmlInput = manifest.toString();
  } else {
    throw new Error(`android-manifest: invalid manifest value passed in: ${manifest}`);
  }
  const indentString = stringTimesN(indentLevel, ' ');

  let formatted = '';
  const regex = /(>)(<)(\/*)/g;
  const xml = xmlInput.replace(regex, `$1${newline}$2$3`);
  let pad = 0;
  xml
    .split(/\r?\n/)
    .map((line: string) => line.trim())
    .forEach((line: string) => {
      let indent = 0;
      if (line.match(/.+<\/\w[^>]*>$/)) {
        indent = 0;
      } else if (line.match(/^<\/\w/)) {
        // Somehow istanbul doesn't see the else case as covered, although it is. Skip it.
        /* istanbul ignore else  */
        if (pad !== 0) {
          pad -= 1;
        }
      } else if (line.match(/^<\w([^>]*[^\/])?>.*$/)) {
        indent = 1;
      } else {
        indent = 0;
      }

      const padding = stringTimesN(pad, indentString);
      formatted += padding + line + newline; // eslint-disable-line prefer-template
      pad += indent;
    });

  return formatted.trim();
}

export async function writeAndroidManifestAsync(
  manifestPath: string,
  manifest: any
): Promise<void> {
  await fs.ensureDir(path.dirname(manifestPath));
  await fs.writeFile(manifestPath, manifest);
}

export async function getProjectAndroidManifestPathAsync(
  projectDir: string
): Promise<string | null> {
  try {
    const shellPath = path.join(projectDir, 'android');
    if ((await fs.stat(shellPath)).isDirectory()) {
      const manifestPath = path.join(shellPath, 'app/src/main/AndroidManifest.xml');
      if ((await fs.stat(manifestPath)).isFile()) {
        return manifestPath;
      }
    }
  } catch (error) {}

  return null;
}

export async function readAsync(manifestPath: string): Promise<Document> {
  const contents = await fs.readFile(manifestPath, { encoding: 'utf8', flag: 'r' });
  const manifest: Document = parseXml(contents);
  return manifest;
}

export async function persistAndroidPermissionsAsync(
  projectDir: string,
  permissions: string[]
): Promise<boolean> {
  const manifestPath = await getProjectAndroidManifestPathAsync(projectDir);
  // The Android Manifest takes priority over the app.json
  if (!manifestPath) {
    return false;
  }
  const manifest = await readAsync(manifestPath);
  removePermissions(manifest);
  const results = ensurePermissions(manifest, permissions);
  if (Object.values(results).reduce((prev, current) => prev && current, true) === false) {
    const failed = Object.keys(results).filter(key => !results[key]);
    throw new Error(
      `Failed to write the following permissions to the native AndroidManifest.xml: ${failed.join(
        ', '
      )}`
    );
  }
  await writeAndroidManifestAsync(manifestPath, manifest);
  return true;
}

// TODO(Bacon): link to resources about required permission prompts
export const UnimodulePermissions: { [key: string]: string } = {
  'android.permission.ACCESS_COARSE_LOCATION': 'ACCESS_COARSE_LOCATION',
  'android.permission.ACCESS_FINE_LOCATION': 'ACCESS_FINE_LOCATION',
  'android.permission.CAMERA': 'CAMERA',
  'android.permission.MANAGE_DOCUMENTS': 'MANAGE_DOCUMENTS',
  'android.permission.READ_CONTACTS': 'READ_CONTACTS',
  'android.permission.READ_CALENDAR': 'READ_CALENDAR',
  'android.permission.WRITE_CALENDAR': 'WRITE_CALENDAR',
  'android.permission.READ_EXTERNAL_STORAGE': 'READ_EXTERNAL_STORAGE',
  'android.permission.READ_PHONE_STATE': 'READ_PHONE_STATE',
  'android.permission.RECORD_AUDIO': 'RECORD_AUDIO',
  'android.permission.USE_FINGERPRINT': 'USE_FINGERPRINT',
  'android.permission.VIBRATE': 'VIBRATE',
  'android.permission.WAKE_LOCK': 'WAKE_LOCK',
  'android.permission.WRITE_EXTERNAL_STORAGE': 'WRITE_EXTERNAL_STORAGE',
  'com.anddoes.launcher.permission.UPDATE_COUNT': 'com.anddoes.launcher.permission.UPDATE_COUNT',
  'com.android.launcher.permission.INSTALL_SHORTCUT':
    'com.android.launcher.permission.INSTALL_SHORTCUT',
  'com.google.android.c2dm.permission.RECEIVE': 'com.google.android.c2dm.permission.RECEIVE',
  'com.google.android.gms.permission.ACTIVITY_RECOGNITION':
    'com.google.android.gms.permission.ACTIVITY_RECOGNITION',
  'com.google.android.providers.gsf.permission.READ_GSERVICES':
    'com.google.android.providers.gsf.permission.READ_GSERVICES',
  'com.htc.launcher.permission.READ_SETTINGS': 'com.htc.launcher.permission.READ_SETTINGS',
  'com.htc.launcher.permission.UPDATE_SHORTCUT': 'com.htc.launcher.permission.UPDATE_SHORTCUT',
  'com.majeur.launcher.permission.UPDATE_BADGE': 'com.majeur.launcher.permission.UPDATE_BADGE',
  'com.sec.android.provider.badge.permission.READ':
    'com.sec.android.provider.badge.permission.READ',
  'com.sec.android.provider.badge.permission.WRITE':
    'com.sec.android.provider.badge.permission.WRITE',
  'com.sonyericsson.home.permission.BROADCAST_BADGE':
    'com.sonyericsson.home.permission.BROADCAST_BADGE',
};
