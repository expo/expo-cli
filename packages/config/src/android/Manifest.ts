import fs from 'fs-extra';
import { Builder, Parser } from 'xml2js';

import { EOL } from 'os';
import path from 'path';

const USES_PERMISSION = 'uses-permission';

type Document = { [key: string]: any };

export function removePermissions(doc: Document, permissionNames?: string[]) {
  const targetNames = permissionNames ? permissionNames.map(ensurePermissionNameFormat) : null;
  const permissions = doc.manifest[USES_PERMISSION] || [];
  let nextPermissions = [];
  for (let attribute of permissions) {
    if (targetNames) {
      const value = attribute['$']['android:name'] || attribute['$']['name'];
      if (!targetNames.includes(value)) {
        nextPermissions.push(attribute);
      }
    }
  }

  doc.manifest[USES_PERMISSION] = nextPermissions;
}

export function addPermission(doc: Document, permissionName: string): void {
  const usesPermissions: { [key: string]: any }[] = doc.manifest[USES_PERMISSION] || [];
  usesPermissions.push({
    $: { 'android:name': permissionName },
  });
  doc.manifest[USES_PERMISSION] = usesPermissions;
}

export function ensurePermissions(
  doc: Document,
  permissionNames: string[]
): { [permission: string]: boolean } {
  const permissions = getPermissions(doc);

  const results: { [permission: string]: boolean } = {};
  for (const permissionName of permissionNames) {
    const targetName = ensurePermissionNameFormat(permissionName);
    if (!permissions.includes(targetName)) {
      addPermission(doc, targetName);
      results[permissionName] = true;
    } else {
      results[permissionName] = false;
    }
  }
  return results;
}

export function ensurePermission(doc: Document, permissionName: string): boolean {
  const permissions = getPermissions(doc);
  const targetName = ensurePermissionNameFormat(permissionName);

  if (!permissions.includes(targetName)) {
    addPermission(doc, targetName);
    return true;
  }
  return false;
}

export function ensurePermissionNameFormat(permissionName: string): string {
  if (permissionName.includes('.')) {
    const com = permissionName.split('.');
    const name = com.pop() as string;
    return [...com, name.toUpperCase()].join('.');
  } else {
    // If shorthand form like `WRITE_CONTACTS` is provided, expand it to `android.permission.WRITE_CONTACTS`.
    return ensurePermissionNameFormat(`android.permission.${permissionName}`);
  }
}

export function getPermissionAttributes(doc: Document): Document[] {
  return doc.manifest[USES_PERMISSION] || [];
}

export function getPermissions(doc: Document): string[] {
  const usesPermissions: { [key: string]: any }[] = doc.manifest[USES_PERMISSION] || [];
  const permissions = usesPermissions.map(permissionObject => {
    return permissionObject['$']['android:name'] || permissionObject['$']['name'];
  });
  return permissions;
}

export function logManifest(doc: Document) {
  const builder = new Builder();
  const xmlInput = builder.buildObject(doc);
  console.log(xmlInput);
}

const stringTimesN = (n: number, char: string) => Array(n + 1).join(char);

export function format(manifest: any, { indentLevel = 2, newline = EOL } = {}): string {
  let xmlInput: string;
  if (typeof manifest === 'string') {
    xmlInput = manifest;
  } else if (manifest.toString) {
    const builder = new Builder({ headless: true });
    xmlInput = builder.buildObject(manifest);
    return xmlInput;
  } else {
    throw new Error(`@expo/android-manifest: invalid manifest value passed in: ${manifest}`);
  }
  const indentString = stringTimesN(indentLevel, ' ');

  let formatted = '';
  const regex = /(>)(<)(\/*)/g;
  const xml = xmlInput.replace(regex, `$1${newline}$2$3`);
  let pad = 0;
  xml
    .split(/\r?\n/)
    .map((line: string) => line.trim())
    .forEach((line: string) => {
      let indent = 0;
      if (line.match(/.+<\/\w[^>]*>$/)) {
        indent = 0;
      } else if (line.match(/^<\/\w/)) {
        // Somehow istanbul doesn't see the else case as covered, although it is. Skip it.
        /* istanbul ignore else  */
        if (pad !== 0) {
          pad -= 1;
        }
      } else if (line.match(/^<\w([^>]*[^\/])?>.*$/)) {
        indent = 1;
      } else {
        indent = 0;
      }

      const padding = stringTimesN(pad, indentString);
      formatted += padding + line + newline; // eslint-disable-line prefer-template
      pad += indent;
    });

  return formatted.trim();
}

export async function writeAndroidManifestAsync(
  manifestPath: string,
  manifest: any
): Promise<void> {
  const manifestXml = new Builder().buildObject(manifest);
  await fs.ensureDir(path.dirname(manifestPath));
  await fs.writeFile(manifestPath, manifestXml);
}

export async function getProjectAndroidManifestPathAsync(
  projectDir: string
): Promise<string | null> {
  try {
    const shellPath = path.join(projectDir, 'android');
    if ((await fs.stat(shellPath)).isDirectory()) {
      const manifestPath = path.join(shellPath, 'app/src/main/AndroidManifest.xml');
      if ((await fs.stat(manifestPath)).isFile()) {
        return manifestPath;
      }
    }
  } catch (error) {}

  return null;
}

export async function readAsync(manifestPath: string): Promise<Document> {
  const contents = await fs.readFile(manifestPath, { encoding: 'utf8', flag: 'r' });
  const parser = new Parser();
  const manifest = parser.parseStringPromise(contents);
  return manifest;
}

export async function persistAndroidPermissionsAsync(
  projectDir: string,
  permissions: string[]
): Promise<boolean> {
  const manifestPath = await getProjectAndroidManifestPathAsync(projectDir);
  // The Android Manifest takes priority over the app.json
  if (!manifestPath) {
    return false;
  }
  const manifest = await readAsync(manifestPath);
  removePermissions(manifest);
  const results = ensurePermissions(manifest, permissions);
  if (Object.values(results).reduce((prev, current) => prev && current, true) === false) {
    const failed = Object.keys(results).filter(key => !results[key]);
    throw new Error(
      `Failed to write the following permissions to the native AndroidManifest.xml: ${failed.join(
        ', '
      )}`
    );
  }

  await writeAndroidManifestAsync(manifestPath, manifest);
  return true;
}

// TODO(Bacon): link to resources about required permission prompts
export const UnimodulePermissions: { [key: string]: string } = {
  'android.permission.READ_INTERNAL_STORAGE': 'READ_INTERNAL_STORAGE',
  'android.permission.ACCESS_COARSE_LOCATION': 'ACCESS_COARSE_LOCATION',
  'android.permission.ACCESS_FINE_LOCATION': 'ACCESS_FINE_LOCATION',
  'android.permission.CAMERA': 'CAMERA',
  'android.permission.MANAGE_DOCUMENTS': 'MANAGE_DOCUMENTS',
  'android.permission.READ_CONTACTS': 'READ_CONTACTS',
  'android.permission.READ_CALENDAR': 'READ_CALENDAR',
  'android.permission.WRITE_CALENDAR': 'WRITE_CALENDAR',
  'android.permission.READ_EXTERNAL_STORAGE': 'READ_EXTERNAL_STORAGE',
  'android.permission.READ_PHONE_STATE': 'READ_PHONE_STATE',
  'android.permission.RECORD_AUDIO': 'RECORD_AUDIO',
  'android.permission.USE_FINGERPRINT': 'USE_FINGERPRINT',
  'android.permission.VIBRATE': 'VIBRATE',
  'android.permission.WAKE_LOCK': 'WAKE_LOCK',
  'android.permission.WRITE_EXTERNAL_STORAGE': 'WRITE_EXTERNAL_STORAGE',
  'com.anddoes.launcher.permission.UPDATE_COUNT': 'com.anddoes.launcher.permission.UPDATE_COUNT',
  'com.android.launcher.permission.INSTALL_SHORTCUT':
    'com.android.launcher.permission.INSTALL_SHORTCUT',
  'com.google.android.c2dm.permission.RECEIVE': 'com.google.android.c2dm.permission.RECEIVE',
  'com.google.android.gms.permission.ACTIVITY_RECOGNITION':
    'com.google.android.gms.permission.ACTIVITY_RECOGNITION',
  'com.google.android.providers.gsf.permission.READ_GSERVICES':
    'com.google.android.providers.gsf.permission.READ_GSERVICES',
  'com.htc.launcher.permission.READ_SETTINGS': 'com.htc.launcher.permission.READ_SETTINGS',
  'com.htc.launcher.permission.UPDATE_SHORTCUT': 'com.htc.launcher.permission.UPDATE_SHORTCUT',
  'com.majeur.launcher.permission.UPDATE_BADGE': 'com.majeur.launcher.permission.UPDATE_BADGE',
  'com.sec.android.provider.badge.permission.READ':
    'com.sec.android.provider.badge.permission.READ',
  'com.sec.android.provider.badge.permission.WRITE':
    'com.sec.android.provider.badge.permission.WRITE',
  'com.sonyericsson.home.permission.BROADCAST_BADGE':
    'com.sonyericsson.home.permission.BROADCAST_BADGE',
};
