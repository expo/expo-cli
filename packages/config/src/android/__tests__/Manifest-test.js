import { dirname, resolve } from 'path';
import fs from 'fs-extra';
import * as Manifest from '../Manifest';
const fixturesPath = resolve(__dirname, 'fixtures');
const manifestPath = resolve(fixturesPath, 'react-native-AndroidManifest.xml');

describe('Permissions', () => {
  it(`adds a new permission`, async () => {
    const manifest = await Manifest.readAsync(manifestPath);
    const didAdd = Manifest.ensurePermission(manifest, 'EXPO_TEST_PERMISSION');
    const permissions = Manifest.getPermissions(manifest);
    expect(didAdd).toBe(true);
    expect(permissions).toContain('android.permission.EXPO_TEST_PERMISSION');
    expect(permissions.length).toBe(2);
  });

  it(`prevents adding a duplicate permission`, async () => {
    const manifest = await Manifest.readAsync(manifestPath);
    const didAdd = Manifest.ensurePermission(manifest, 'INTERNET');
    const permissions = Manifest.getPermissions(manifest);
    expect(didAdd).toBe(false);
    expect(permissions).toContain('android.permission.INTERNET');
    expect(permissions.length).toBe(1);
  });

  it(`ensures multiple permissions`, async () => {
    const manifest = await Manifest.readAsync(manifestPath);
    const permissionsToAdd = ['VALUE_1', 'VALUE_2'];
    const results = Manifest.ensurePermissions(manifest, permissionsToAdd);
    expect(results).toMatchSnapshot();
    expect(Object.values(results)).toStrictEqual([true, true]);

    expect(Manifest.getPermissions(manifest).length).toBe(3);
  });

  it(`removes permissions by name`, async () => {
    const manifest = await Manifest.readAsync(manifestPath);
    expect(Manifest.ensurePermission(manifest, 'VALUE_TO_REMOVE_1')).toBe(true);
    expect(Manifest.ensurePermission(manifest, 'VALUE_TO_REMOVE_2')).toBe(true);
    expect(Manifest.getPermissions(manifest).length).toBe(3);

    Manifest.removePermissions(manifest, ['VALUE_TO_REMOVE_1', 'VALUE_TO_REMOVE_2']);
    expect(Manifest.getPermissions(manifest).length).toBe(1);
  });

  it(`removes all permissions`, async () => {
    const manifest = await Manifest.readAsync(manifestPath);
    expect(Manifest.ensurePermission(manifest, 'VALUE_TO_REMOVE_1')).toBe(true);
    expect(Manifest.ensurePermission(manifest, 'VALUE_TO_REMOVE_2')).toBe(true);
    expect(Manifest.getPermissions(manifest).length).toBe(3);

    Manifest.removePermissions(manifest);

    expect(Manifest.getPermissions(manifest).length).toBe(0);
  });

  it(`can write with a pretty format`, async () => {
    const manifest = await Manifest.readAsync(manifestPath);
    expect(Manifest.ensurePermission(manifest, 'NEW_PERMISSION_1')).toBe(true);
    expect(Manifest.ensurePermission(manifest, 'NEW_PERMISSION_2')).toBe(true);
    expect(Manifest.getPermissions(manifest).length).toBe(3);

    expect(Manifest.format(manifest)).toMatchSnapshot();
  });

  describe('E2E', () => {
    beforeEach(async () => {
      await fs.remove(resolve(fixturesPath, 'android'));
    });
    afterAll(async () => {
      await fs.remove(resolve(fixturesPath, 'android'));
    });

    const appManifestPath = resolve(fixturesPath, 'android/app/src/main/AndroidManifest.xml');

    it(`replaces and persists permissions`, async () => {
      await fs.ensureDir(dirname(appManifestPath));
      await fs.copyFile(manifestPath, appManifestPath);

      const permissionsToPersist = [
        'CAMERA',
        'android.permission.WRITE_CALENDAR',
        'com.anddoes.launcher.permission.UPDATE_COUNT',
        'com.sonyericsson.home.permission.BROADCAST_BADGE',
      ];
      expect(
        await Manifest.persistAndroidPermissionsAsync(fixturesPath, permissionsToPersist)
      ).toBe(true);

      const manifest = await Manifest.readAsync(appManifestPath);
      expect(Manifest.getPermissions(manifest)).toStrictEqual(
        permissionsToPersist.map(Manifest.ensurePermissionNameFormat)
      );
    });

    it(`returns false when the native project cannot be found`, async () => {
      expect(await Manifest.persistAndroidPermissionsAsync(fixturesPath, [])).toBe(false);
    });
  });
});
