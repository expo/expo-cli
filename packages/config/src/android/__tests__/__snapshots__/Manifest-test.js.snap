// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Permissions can write with a pretty format 1`] = `
"<manifest xmlns:android=\\"http://schemas.android.com/apk/res/android\\" package=\\"com.expo.mycoolapp\\">
  <uses-permission android:name=\\"android.permission.INTERNET\\"/>
  <uses-permission android:name=\\"android.permission.NEW_PERMISSION_1\\"/>
  <uses-permission android:name=\\"android.permission.NEW_PERMISSION_2\\"/>
  <application android:name=\\".MainApplication\\" android:label=\\"@string/app_name\\" android:icon=\\"@mipmap/ic_launcher\\" android:roundIcon=\\"@mipmap/ic_launcher_round\\" android:allowBackup=\\"false\\" android:theme=\\"@style/AppTheme\\">
    <activity android:name=\\".MainActivity\\" android:label=\\"@string/app_name\\" android:configChanges=\\"keyboard|keyboardHidden|orientation|screenSize\\" android:windowSoftInputMode=\\"adjustResize\\">
      <intent-filter>
        <action android:name=\\"android.intent.action.MAIN\\"/>
        <category android:name=\\"android.intent.category.LAUNCHER\\"/>
      </intent-filter>
    </activity>
    <activity android:name=\\"com.facebook.react.devsupport.DevSettingsActivity\\"/>
  </application>
</manifest>"
`;

exports[`Permissions ensures multiple permissions 1`] = `
Object {
  "VALUE_1": true,
  "VALUE_2": true,
}
`;
